overlord/snapstate/refresh.go

```diff
@@ -0,0 +1,214 @@
+	// Grab per-snap lock to prevent new processes from starting. This is
+	// sufficient to perform the check, even though individual processes
+	// may fork or exit, we will have per-security-tag information about
+	// what is running.
+	lock, err := snaplock.OpenLock(info.SnapName())
+	if err != nil {
+		return err
+	}
+	defer lock.Close()
+	if err := lock.Lock(); err != nil {
+		return err
+	}
+
+	var busyAppNames []string
+	var busyHookNames []string
+	var busyPIDs []int
+
+	canHookRunDuringRefresh := func(hook *snap.HookInfo) bool {
+		return false
+	}
+
+	for name, app := range info.Apps {
+		if canAppRunDuringRefresh(app) {
+			continue
+		}
+		PIDs, err := pidsOfSecurityTag(app.SecurityTag())
+		if err != nil {
+			return err
+		}
+		if len(PIDs) > 0 {
+			busyAppNames = append(busyAppNames, name)
+			busyPIDs = append(busyPIDs, PIDs...)
+		}
+	}
+
+	for name, hook := range info.Hooks {
+		if canHookRunDuringRefresh(hook) {
+			continue
+		}
+		PIDs, err := pidsOfSecurityTag(hook.SecurityTag())
+		if err != nil {
+			return err
+		}
+		if len(PIDs) > 0 {
+			busyHookNames = append(busyHookNames, name)
+			busyPIDs = append(busyPIDs, PIDs...)
+		}
+	}
+	if len(busyAppNames) == 0 && len(busyHookNames) == 0 {
+		return nil
+	}
+	sort.Strings(busyAppNames)
+	sort.Strings(busyHookNames)
+	sort.Ints(busyPIDs)
+	return &BusySnapError{
+		snapName:      info.SnapName(),
+		busyAppNames:  busyAppNames,
+		busyHookNames: busyHookNames,
+		pids:          busyPIDs,
+	}
+}
+
+// SoftNothingRunningRefreshCheck looks if there are at most only service processes alive.
+//
+// The check is designed to run early in the refresh pipeline. Before
+// downloading or stopping services for the update, we can check that only
+// services are running, that is, that no non-service apps or hooks are
+// currently running.
+//
+// Since services are stopped during the update this provides a good early
+// precondition check.  The check is also deliberately racy, both at the level
+// of processes forking and exiting and at the level of snap-confine launching
+// new commands. The hard check needs to be synchronized but the soft check
+// doesn't require this since it would serve no purpose. After the soft check
+// passes the user is free to start snap applications and block the hard check.
+func SoftNothingRunningRefreshCheck(info *snap.Info) error {
+	return genericRefreshCheck(info, func(app *snap.AppInfo) bool {
+		return app.IsService()
+	})
+}
+
+// HardNothingRunningRefreshCheck looks if there are any processes alive.
+//
+// The check is designed to run late in the refresh pipeline, after stopping
+// snap services. At this point services should be stopped, hooks should no
+// longer run, and applications should be barred from running externally (e.g.
+// by grabbing the per-snap lock around that phase of the update).
+//
+// The check looks at the set of PIDs in the freezer cgroup associated with a
```

Comment:
the comment about the freezer group seems untrue, we are only looking at pid groups, no?

Reply:
Yes, the comment needs to be updated now. Thank you!